pkg/controller/rolloutblock/rolloutblock_controller.go

```diff
@@ -0,0 +1,285 @@
+)
+
+// Controller is a Kubernetes controller that updates rolloutblock objects.
+// The RolloutBlock objects' main objective is to block application and releases rollout during
+// an outage.
+//
+// RolloutBlock Controller has one primary workqueues: a rolloutblock updater
+// object queue.
+type Controller struct {
+	shipperClientset clientset.Interface
+	recorder         record.EventRecorder
+
+	applicationLister  shipperlisters.ApplicationLister
+	applicationsSynced cache.InformerSynced
+
+	releaseLister  shipperlisters.ReleaseLister
+	releasesSynced cache.InformerSynced
+
+	rolloutBlockLister shipperlisters.RolloutBlockLister
+	rolloutBlockSynced cache.InformerSynced
+
+	rolloutblockWorkqueue workqueue.RateLimitingInterface
+}
+
+// NewController returns a new RolloutBlock controller.
+func NewController(
+	shipperClientset clientset.Interface,
+	informerFactory shipperinformers.SharedInformerFactory,
+	recorder record.EventRecorder,
+) *Controller {
+	applicationInformer := informerFactory.Shipper().V1alpha1().Applications()
+	releaseInformer := informerFactory.Shipper().V1alpha1().Releases()
+	rolloutBlockInformer := informerFactory.Shipper().V1alpha1().RolloutBlocks()
+
+	glog.Info("Building a RolloutBlock controller")
+
+	controller := &Controller{
+		recorder:         recorder,
+		shipperClientset: shipperClientset,
+
+		applicationLister:  applicationInformer.Lister(),
+		applicationsSynced: applicationInformer.Informer().HasSynced,
+
+		releaseLister:  releaseInformer.Lister(),
+		releasesSynced: releaseInformer.Informer().HasSynced,
+
+		rolloutBlockLister: rolloutBlockInformer.Lister(),
+		rolloutBlockSynced: rolloutBlockInformer.Informer().HasSynced,
+
+		rolloutblockWorkqueue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(),
+			"rolloutblock_controller_rolloutblocks",
+		),
+	}
+
+	glog.Info("Setting up event handlers")
+
+	releaseInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: controller.enqueueRelease,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				controller.enqueueRelease(newObj)
+			},
+			DeleteFunc: controller.enqueueRelease,
+		})
+
+	applicationInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: controller.enqueueApplication,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				controller.enqueueApplication(newObj)
+			},
+			DeleteFunc: controller.enqueueApplication,
+		})
+
+	rolloutBlockInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: controller.onAddRolloutBlock,
+		})
+
+	return controller
+}
+
+func (c *Controller) enqueueRelease(obj interface{}) {
+	rel, ok := obj.(*shipper.Release)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("not a shipper.Release: %#v", obj))
+		return
+	}
+
+	overrideRB, _ := rel.GetAnnotations()[shipper.RolloutBlocksOverrideAnnotation]
+	overrideRBs := rolloutBlockOverride.NewOverride(overrideRB)
+	for rbKey := range overrideRBs {
```

Comment:
I wonder if there is a specific reason for plural enqueueing, not singular stringified?